core-compose/src/main/java/com/squareup/workflow/ui/compose/WorkflowContainer.kt

```diff
@@ -0,0 +1,323 @@
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, viewEnvironment, Unit, onOutput, modifier, diagnosticListener)
+}
+
+/**
+ * Render a [Workflow]'s renderings.
+ *
+ * When this function is first composed it will start a new runtime. This runtime will be restarted
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param viewEnvironment The [ViewEnvironment] used to show the [ComposeRendering]s emitted by
+ * the workflow.
+ * @param props The props to render the root workflow with. If this value changes between calls,
+ * the workflow runtime will re-render with the new props.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ */
+@Direct
+@Composable fun <PropsT> WorkflowContainer(
+  workflow: Workflow<PropsT, Nothing, ComposeRendering>,
+  viewEnvironment: ViewEnvironment,
+  props: PropsT,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, viewEnvironment, props, {}, modifier, diagnosticListener)
+}
+
+/**
+ * Render a [Workflow]'s renderings.
+ *
+ * When this function is first composed it will start a new runtime. This runtime will be restarted
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param viewEnvironment The [ViewEnvironment] used to show the [ComposeRendering]s emitted by
+ * the workflow.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ */
+@Direct
+@Composable fun WorkflowContainer(
+  workflow: Workflow<Unit, Nothing, ComposeRendering>,
+  viewEnvironment: ViewEnvironment,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, viewEnvironment, Unit, {}, modifier, diagnosticListener)
+}
+
+@Composable private fun <PropsT, OutputT : Any, RenderingT> observeWorkflow(
+  @Pivotal workflow: Workflow<PropsT, OutputT, RenderingT>,
+  props: PropsT,
+  onOutput: (OutputT) -> Unit,
+  @Pivotal coroutineContext: CoroutineContext,
+  @Pivotal diagnosticListener: WorkflowDiagnosticListener? = null
+): RenderingT {
+  // This can be a StateFlow once coroutines is upgraded to 1.3.6.
+  val propsChannel = remember { Channel<PropsT>(capacity = CONFLATED) }
+  propsChannel.offer(props)
+
+  // Need a mutable holder for onOutput so the outputs subscriber created in the onActive block
+  // will always be able to see the latest value.
+  val outputCallback = remember { OutputCallback(onOutput) }
+  outputCallback.onOutput = onOutput
+
+  val renderingState = state<RenderingT?> { null }
+  val snapshotState = savedInstanceState(saver = SnapshotSaver) { null }
```

Comment:
Should this be accepted as a parameter? Would make caller responsible for saving this to the instance state themselves.

Reply:
I don't think there's a use case. `setContentWorkflow` doesn't even provide this API. Not gonna do it.